project/project.go

```diff
@@ -1,70 +1,70 @@
 package project
 
 import (
+	"fmt"
 	"io/ioutil"
 	"path/filepath"
 
 	"github.com/BurntSushi/toml"
 	"github.com/pkg/errors"
 
-	"github.com/buildpacks/pack/internal/dist"
-)
-
-type Script struct {
-	API    string `toml:"api"`
-	Inline string `toml:"inline"`
-	Shell  string `toml:"shell"`
-}
-
-type Buildpack struct {
-	ID      string `toml:"id"`
-	Version string `toml:"version"`
-	URI     string `toml:"uri"`
-	Script  Script `toml:"script"`
-}
+	"github.com/buildpacks/lifecycle/api"
 
-type EnvVar struct {
-	Name  string `toml:"name"`
-	Value string `toml:"value"`
-}
+	"github.com/buildpacks/pack/project/common"
+	v01 "github.com/buildpacks/pack/project/v01"
+	v02 "github.com/buildpacks/pack/project/v02"
+)
 
-type Build struct {
-	Include    []string    `toml:"include"`
-	Exclude    []string    `toml:"exclude"`
-	Buildpacks []Buildpack `toml:"buildpacks"`
-	Env        []EnvVar    `toml:"env"`
-	Builder    string      `toml:"builder"`
+var supportedSchemas = map[string]common.ProjectDescriptorSchema{
+	"0.1": v01.Descriptor{},
+	"0.2": v02.Descriptor{},
 }
 
 type Project struct {
-	Name     string         `toml:"name"`
-	Licenses []dist.License `toml:"licenses"`
+	Version string `toml:"schema-version"`
 }
-
-type Descriptor struct {
-	Project  Project                `toml:"project"`
-	Build    Build                  `toml:"build"`
-	Metadata map[string]interface{} `toml:"metadata"`
+type VersionDescriptor struct {
+	Project Project `toml:"_"`
 }
 
-func ReadProjectDescriptor(pathToFile string) (Descriptor, error) {
+func ReadProjectDescriptor(pathToFile string) (common.Descriptor, error) {
```

Comment:
I'm personally finding the self-updating structs a little awkward.

```suggestion
func ReadProjectDescriptor(pathToFile string) (common.Descriptor, error) {
	pathToFile = filepath.Clean(pathToFile)

	var versionDescriptor struct {
		Project struct {
			Version string `toml:"schema-version"`
		} `toml:"_"`
	}

	_, err := toml.DecodeFile(pathToFile, &versionDescriptor)
	if err != nil {
		return common.Descriptor{}, errors.Wrapf(err, "parsing schema version")
	}

	switch versionDescriptor.Project.Version {
	// _.schema-version does not exist in 0.1
	case "0.1", "":
		descriptor, err := v01.NewDescriptor(pathToFile)
		if err != nil {
			return common.Descriptor{}, err
		}

		return descriptor, validate(descriptor)
	case "0.2":
		descriptor, err := v02.NewDescriptor(pathToFile)
		if err != nil {
			return common.Descriptor{}, err
		}

		return descriptor, validate(descriptor)
	default:
		return common.Descriptor{}, fmt.Errorf("unknown project descriptor schema version %s", versionDescriptor.Project.Version)
	}
}
```

Reply:
I like this suggestion a lot! I think it looks a lot cleaner and less awkward. The only possible downside I could see is if we have lots of versions in the future, the switch statement could get a little large and redundant. What are your thoughts on the following:

```
var NewDescriptor = map[string]common.NewDescriptorType{
	"0.1": v01.NewDescriptor,
	"0.2": v02.NewDescriptor,
}

func ReadProjectDescriptor(pathToFile string) (common.Descriptor, error) {
	projectTomlContents, err := ioutil.ReadFile(filepath.Clean(pathToFile))
	if err != nil {
		return common.Descriptor{}, err
	}

	var versionDescriptor struct {
		Project struct {
			Version string `toml:"schema-version"`
		} `toml:"_"`
	}

	_, err = toml.Decode(string(projectTomlContents), &versionDescriptor)
	if err != nil {
		return common.Descriptor{}, errors.Wrapf(err, "parsing schema version")
	}

	version := versionDescriptor.Project.Version
	if version == "" {
		version = "0.1"
	}

	var descriptor common.Descriptor

	if _, ok := NewDescriptor[version]; !ok {
		return common.Descriptor{}, fmt.Errorf("unknown project descriptor schema version %s", version)
	}

	descriptor, err = NewDescriptor[version](pathToFile)
	if err != nil {
		return common.Descriptor{}, err
	}

	return descriptor, validate(descriptor)

}
```

I'm also totally happy to land on your suggestion; I do think it's the cleanest solution for the present. No strong feelings here, just wanted to put out one more thought 😄

Reply:
I don't know why I didn't get a notification. I'm so sorry for missing this! 🙇🏾 

Ultimately, I think it's a style preference which is why I have no problem with what you've presented. In my personal projects, large and redundant `switch` statements are very common; But I understand that it's not for everyone!

Reply:
Since we slacked a bit here @YousefHaggyHeroku I'm going through the process of creating a polish PR for this so we can get this in ASAP. I'm like and therefore taking your latest suggestion into the follow up PR. I'll also take care of the merge conflicts for you.

Reply:
Sounds good! Thanks for making that PR